### PR TITLE
Add single exception catch to study optimize

### PR DESCRIPTION
**Pull request by gen740**
## Motivation
Related to #3230, support single Exception catch for user friendliness.

Currently if you want to pass `SomeException` as `catch` to `optuna.Study.optimize`, you should write like this
```
study.optimize(objective, catch=(SomeException,), n_trials=100)
```

It is intuitive to write like this
```
study.optimize(objective, catch=SomeException, n_trials=100)
```

## Description of the changes
In `Study.optimize`, catch value will passed to `_optimize` if catch is not a tuple.

There is a way to change all functions which would use `catch` (like `_optimize`, `_optimize_sequential`, and so). But for internal usage, `Tuple[Type[Exception], ...]` would be enough.

### CODE REVIEW

**Comment by contramundum53** [comment]
Thanks for the PR!
@HideakiImamura Could you review this PR?

**Comment by contramundum53** [comment]
@gen740 How about extending the type to `Union[Container[Type[Exception], ...], Type[Exception]]` as mentioned in the original issue?

File: optuna/study/study.py [hunk 1]
```diff
@@ -426,7 +426,7 @@ def objective(trial):
             n_trials=n_trials,
             timeout=timeout,
             n_jobs=n_jobs,
-            catch=catch,
+            catch=catch if isinstance(catch, tuple) else (catch,),
```
Comment (nzw0301) [inline optuna/study/study.py:429]: `tuple` should take iterable according to its definition. 

```suggestion
            catch=tuple(catch) if isinstance(catch, Iterable) else (catch,),
```

File: optuna/study/study.py [hunk 2]
```diff
@@ -322,7 +322,7 @@ def optimize(
         n_trials: Optional[int] = None,
         timeout: Optional[float] = None,
         n_jobs: int = 1,
-        catch: Tuple[Type[Exception], ...] = (),
+        catch: Union[Tuple[Type[Exception], ...], Type[Exception]] = (),
```
Comment (nzw0301) [inline optuna/study/study.py:325]: Similarly, this part might be relaxed from `Tuple` to `Iterable`.
```suggestion
        catch: Union[Iterable[Type[Exception]], Type[Exception]] = (),
```

**Review by nzw0301 — COMMENTED** [review]
In the current code, if `catch` takes a `list` of `Exception` that looks more natural according to users' opinions in https://github.com/optuna/optuna/issues/3230, the error message looks confusing to fix the code as follows:

```python
import optuna

def objective(trial):
    raise ValueError()
    return trial.suggest_float("x", -1, 1, step=0.1)


study = optuna.create_study(sampler=optuna.samplers.TPESampler(seed=2))
study.optimize(objective, n_trials=20, catch=[ValueError])
```

The error message is as follows:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [1], in <cell line: 9>()
      5     return trial.suggest_float("x", -1, 1, step=0.1)
      8 study = optuna.create_study(sampler=optuna.samplers.TPESampler(seed=2))
----> 9 study.optimize(objective, n_trials=20, catch=[ValueError])

File ~/Documents/optuna/optuna/study/study.py:423, in Study.optimize(self, func, n_trials, timeout, n_jobs, catch, callbacks, gc_after_trial, show_progress_bar)
    319 def optimize(
    320     self,
    321     func: ObjectiveFuncType,
   (...)
    328     show_progress_bar: bool = False,
    329 ) -> None:
    330     """Optimize an objective function.
    331 
    332     Optimization is done by choosing a suitable set of hyperparameter values from a given
   (...)
    420             If nested invocation of this method occurs.
    421     """
--> 423     _optimize(
    424         study=self,
    425         func=func,
    426         n_trials=n_trials,
    427         timeout=timeout,
    428         n_jobs=n_jobs,
    429         catch=catch if isinstance(catch, tuple) else (catch,),
    430         callbacks=callbacks,
    431         gc_after_trial=gc_after_trial,
    432         show_progress_bar=show_progress_bar,
    433     )

File ~/Documents/optuna/optuna/study/_optimize.py:66, in _optimize(study, func, n_trials, timeout, n_jobs, catch, callbacks, gc_after_trial, show_progress_bar)
     64 try:
     65     if n_jobs == 1:
---> 66         _optimize_sequential(
     67             study,
     68             func,
     69             n_trials,
     70             timeout,
     71             catch,
     72             callbacks,
     73             gc_after_trial,
     74             reseed_sampler_rng=False,
     75             time_start=None,
     76             progress_bar=progress_bar,
     77         )
     78     else:
     79         if n_jobs == -1:

File ~/Documents/optuna/optuna/study/_optimize.py:163, in _optimize_sequential(study, func, n_trials, timeout, catch, callbacks, gc_after_trial, reseed_sampler_rng, time_start, progress_bar)
    160         break
    162 try:
--> 163     frozen_trial = _run_trial(study, func, catch)
    164 finally:
    165     # The following line mitigates memory problems that can be occurred in some
    166     # environments (e.g., services that use computing containers such as CircleCI).
    167     # Please refer to the following PR for further details:
    168     # https://github.com/optuna/optuna/pull/325.
    169     if gc_after_trial:

File ~/Documents/optuna/optuna/study/_optimize.py:248, in _run_trial(study, func, catch)
    242     else:
    243         assert False, "Should not reach."
    245 if (
    246     frozen_trial.state == TrialState.FAIL
    247     and func_err is not None
--> 248     and not isinstance(func_err, catch)
    249 ):
    250     raise func_err
    251 return frozen_trial

TypeError: isinstance() arg 2 must be a type or tuple of types
```

I'd like to give a few suggests to relax this by using `Itetable` that is an argument type of `Tuple` since `_optimize` checks whether or not `catch` is a tuple. 

In addition, [`chainermn`'s `optimze` function has also `catch` argument](https://github.com/optuna/optuna/blob/6ad71fc4224e9652320195e37d22ebcd1f9dbf2f/optuna/integration/chainermn.py#L105), so it would be more consistent with the current change.

**Review by nzw0301 — APPROVED** [review]
Brilliant! LGTM!

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. The change looks good to me. How about adding some test cases to verify a single exception and not tuple exceptions?

**Review by contramundum53 — APPROVED** [review]
LGTM!

**Comment by gen740** [comment]
I guess, python built-in (and Most used) Iterable types are, 'tuple', 'list', and 'set'. (Of course 'dictionary' is also Iterable, but no one will want to use 'dictionary' as a 'catch' parameter.)

I would write a test for these 'Exception', 'tuple', 'list', and 'set' cases, and I think this is enough for this feature. Because, except for 'Exception', Iterable and tuple convertible types are trivially acceptable.

File: tests/study_tests/test_study.py [hunk 3]
```diff
@@ -237,13 +237,11 @@ def test_optimize_with_catch(storage_mode: str) -> None:
         assert all(trial.state == TrialState.FAIL for trial in study.trials)
 
 
-@pytest.mark.parametrize("catch", [[], [Exception], None, 1])
-def test_optimize_with_catch_invalid_type(catch: Any) -> None:
```
[the remark this answers is not in the document]
Comment (gen740) [reply tests/study_tests/test_study.py:241]: Oh! I see. Thank you!

**Review by HideakiImamura — APPROVED** [review]
Looks very good.

**Pull request merged**